rfcs/0000-multi-server-course-content.md

```diff
@@ -0,0 +1,58 @@
+* Viewing or grading a question
+  * Custom elements chunk
+  * Question chunk
+  * `serverFilesCourse` chunk
+  * Appropriate `serverFilesCourseInstance` chunk
+* Serving a file from `clientFilesCourse`
+  * `clientFilesCourse` chunk
+
+Note that in many cases, the chunk will have already been loaded to disk for a previous request, and so it can be reused. In fact, this is what makes this method so attractive: it allows us to amortize the cost of loading a course's content over the course of many requests. Files are loaded on-demand and are aggressively cached locally, meaning that requests will only be slightly slower (on the order of hundreds of milliseconds) when they requires course chunks that don't yet exist locally on disk.
+
+## Syncing
+
+In order to sync course content, we'll still need entire course repos to exist *somewhere* on disk. To avoid the overhead of needing to clone an entire course repo for each sync, we'll dedicate a single PrairieLearn instance solely to syncing operations. This instance will be like the current PrairieLearn server in that it will contain all course repos on disk. We'll configure a load balancer to route all syncing requests to that instance. That instance will then sync metadata to the database and upload the course's chunks as appropriate. Note that this means we don't need a separate service to handle syncing - when running locally, all syncing requests will go to the same server as usual. And servers running in production don't even need to be aware of whether or not they're the syncing instance. This keeps the overall design simple.
+
+When a course instance is synced, we update its git hash in the database. This hash will serve as a unique version identifier for the course's contents. The next time someone tries to access something from the course, we'll check if we already have a working directory for that particular version. If not, we can create one that will store only files for the particular revision, like `/courses/my_course_id/234fae...[hash]...9823efde/`. We'll then start loading files again to this new directory so as not to disrupt any question renders that might be in progress. We can then have a background task that periodically deletes old versions.
+
+## Local storage exhaustion
+
+Over time, the state of the world will trend towards all course files for all courses existing on disk. Ideally, autoscaled instances are running on smaller, cheaper machines with limited resources. This means that over time, we're likely to run out of disk space. To prevent this (or if this someday becomes a problem), we can build an LRU cache on top of this system that will delete old content that hasn't been used in a while.
+
+# Potential future optimizations
+
+* The hash of a chunk's content could be stored and incorporated into its ID; this could be used to optimize syncs to only upload changes chunks. And for servers, they would only need to load a chunk again if its contents have changed.
```

Comment:
Ah, ok, now I understand my earlier confusion. I was thinking that we'd do this optimization from the start. Otherwise any tiny change to any part of a course would require all servers to re-download all questions for that course. I've seen instructors updating access rules as a remote exam is starting (typical case is to add an extra-time accommodation for a student at the last minute), and this happens exactly when the servers are all under high load.

I think we need two DB tables:

Table `chunks` to store which chunks are currently available in S3, with columns:
* `course_id`
* `type` - can be `serverFilesCourse`, `elements`, `question`, etc.
* `question_id` - this is NULL unless `type = 'question'`
* `course_instance_id` - this is NULL unless `type = 'clientFilesCourseInstance'`
* `hash` - the currently active hash for this chunk

Table `instance_chunks` to store which chunk versions are on each machine, with columns: 
* `instance_id` - unique identifier for each machine (we already have this as `config.instanceId`)
* `chunk_id`
* `hash` - this is the hash that we have on local disk for this chunk on this instance

When we want to serve a question route, we'd figure out whether we need new chunks with something like

```sql
SELECT
    chunks.*
FROM
    chunks
    LEFT JOIN instance_chunks ON (chunks.id = instance_chunks.chunk_id)
WHERE
    course_id = $course_id
    AND instance_id = $instance_id
    AND (
        type = 'serverFilesCourse'
        OR type = 'elements'
        OR (type = 'question' and question_id = $question_id)
    )
    AND instance_chunks.hash != chunks.hash;
```

We could do this as part of `middlewares/selectAndAuthzInstanceQuestion.sql`, for example, (just like the `file_list` CTE in that file) so that it will add negligible overhead and no each DB round-trips. We can store the list of returned chunks in `res.locals.chunks_needed` and have a middleware that checks this variable and blocks until we've got any needed chunks downloaded.

Reply:
Sorry, I was definitely a bit hand-wavey with how chunks would be stored and identified! Agreed that it makes sense to do this optimization from the start; I can add that explicitly to this RFC.

Storing the chunks used by a particular instance in the DB seems pretty heavyweight, especially considering that we're intending to trend towards many smaller, more ephemeral instances. I know that the DB is the one bit of our stack that will handle anything we can throw at it, but I'd rather not bloat it any more than we have to 🙂. I think just having each instance keep an in-memory cache of which files are on its particular disk would be sufficient. If we can't find it in the in-memory cache, we can fall back to a slightly-slower `stat` to see if it's on disk. If it is, we update the in-memory cache; if it isn't, we go to S3 for the chunk.

This should keep the average case (chunk is on disk already) as fast as possible (in-memory cache is orders of magnitude faster than hitting the disk), while also keeping less common cases (server crashes and restarts; files are still on disk, and a `stat` is still faster than a fresh download from S3.

Reply:
I agree that in-memory (and on-disk) storage of local chunk state is fine. Although it does make it harder for a central process to generate statistics 😄